druid-shell/src/platform/x11/application.rs

```diff
@@ -404,4 +466,57 @@ impl Application {
         log::warn!("Application::get_locale is currently unimplemented for X11 platforms. (defaulting to en-US)");
         "en-US".into()
     }
+
+    pub(crate) fn idle_pipe(&self) -> RawFd {
+        self.idle_write
+    }
+}
+
+/// Returns when there is an event ready to read from `conn`, or there is data ready to read from
+/// `idle` and the `timeout` has passed.
+// This was taken, with minor modifications, from the xclock_utc example in the x11rb crate.
+// https://raw.githubusercontent.com/psychon/x11rb/a6bd1453fd8e931394b9b1f2185fad48b7cca5fe/examples/xclock_utc.rs
+fn poll_with_timeout(conn: &Rc<XCBConnection>, idle: RawFd, timeout: Instant) -> Result<(), Error> {
+    use nix::poll::{poll, PollFd, PollFlags};
+    use std::os::raw::c_int;
+    use std::os::unix::io::AsRawFd;
+
+    let fd = conn.as_raw_fd();
+    let mut poll_fds = [
+        PollFd::new(fd, PollFlags::POLLIN),
+        PollFd::new(idle, PollFlags::POLLIN),
+    ];
+
+    // We start with no timeout in the poll call. If we get something from the idle handler, we'll
+    // start setting one.
+    let mut poll_timeout = -1;
+    loop {
+        let readable = |p: &PollFd| {
+            p.revents()
+                .unwrap_or_else(PollFlags::empty)
```

Comment:
By the way, I have no idea why this `unwrap_or_else` is necessary, but I know that the x11rb example also has it...

Reply:
I'm not sure I understand this comment... we have to do *something* in the unlikely event that `poll (2)` sets `revents` to something invalid.

Reply:
Yeah, this is not really meant as a comment to you. What I mean is: Why can't `nix` do this "something" for us? Is there any other sensible option than `unwrap_or_else(PollFlags::empty)`?